MINOR: Cache topic resolution in TopicIds set

squah-confluent opened the pull request:
Looking up topics in a TopicsImage is relatively slow. Cache the results
in TopicIds to improve assignor performance. In benchmarks, we see a
noticeable improvement in performance in the heterogeneous case.

Before
```
Benchmark                                       (assignmentType)  (assignorType)  (isRackAware)  (memberCount)  (partitionsToMemberRatio)  (subscriptionType)  (topicCount)  Mode  Cnt    Score   Error  Units
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           RANGE          false          10000                         10         HOMOGENEOUS          1000  avgt    5   36.400 ± 3.004  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           RANGE          false          10000                         10       HETEROGENEOUS          1000  avgt    5  158.340 ± 0.825  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL         UNIFORM          false          10000                         10         HOMOGENEOUS          1000  avgt    5    1.329 ± 0.041  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL         UNIFORM          false          10000                         10       HETEROGENEOUS          1000  avgt    5  382.901 ± 6.203  ms/op
```

After
```
Benchmark                                       (assignmentType)  (assignorType)  (isRackAware)  (memberCount)  (partitionsToMemberRatio)  (subscriptionType)  (topicCount)  Mode  Cnt    Score   Error  Units
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           RANGE          false          10000                         10         HOMOGENEOUS          1000  avgt    5   36.465 ± 1.954  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL           RANGE          false          10000                         10       HETEROGENEOUS          1000  avgt    5  114.043 ± 1.424  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL         UNIFORM          false          10000                         10         HOMOGENEOUS          1000  avgt    5    1.454 ± 0.019  ms/op
ServerSideAssignorBenchmark.doAssignment             INCREMENTAL         UNIFORM          false          10000                         10       HETEROGENEOUS          1000  avgt    5  342.840 ± 2.744  ms/op
```

---

Based heavily on https://github.com/apache/kafka/pull/16527.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dajac committed:
MINOR: Cache topic resolution in TopicIds set

Looking up topics in a TopicsImage is relatively slow. Cache the results
in TopicIds to improve assignor performance.

FrankYang0529 committed:
MINOR: fix failed cases in FeatureCommandTest (#17287)

Reviewers: David Arthur <mumrah@gmail.com>

FrankYang0529 committed:
KAFKA-16792: Enable consumer unit tests that fail to fetch offsets only for new consumer with poll(0) (#16982)

Reviewers: Lianet Magrans <lmagrans@confluent.io>, TaiJuWu <tjwu1217@gmail.com>, Kirk True <ktrue@confluent.io>, TengYao Chi <kitingiao@gmail.com>

jolshan committed:
KAFKA-14562 [1/2]: Implement epoch bump after every transaction (#16719)

Implement server side changes for epoch bump but keep EndTxn as an unstable API until the client side changes are implemented. EndTxnResponse will return the producer ID and epoch for the transaction. Introduces new tagged fields to the TransactionLogValue to persist the clientTransactionVersion, previousProducerId, and nextProducerId to the log so that the state can be reloaded. See KIP-890 for more details.

Small updates to naming of lastProducerId -> PreviousProducerId. Also cleans up the many TransactionMetadata constructors.

Reviewers: Artem Livshits <alivshits@confluent.io>, David Jacot <djacot@confluent.io>

ahuang98 committed:
KAFKA-17608, KAFKA-17604, KAFKA-16963; KRaft controller crashes when active controller is removed (#17146)

This change fixes a few issues.

KAFKA-17608; KRaft controller crashes when active controller is removed
When a control batch is committed, the quorum controller currently increases the last stable offset but fails to create a snapshot for that offset. This causes an issue if the quorum controller renounces and needs to revert to that offset (which has no snapshot present). Since the control batches are no-ops for the quorum controller, it does not need to update its offsets for control records. We skip handle commit logic for control batches.

KAFKA-17604; Describe quorum output missing added voters endpoints
Describe quorum output will miss endpoints of voters which were added via AddRaftVoter. This is due to a bug in LeaderState's updateVoterAndObserverStates which will pull replica state from observer states map (which does not include endpoints). The fix is to populate endpoints from the lastVoterSet passed into the method.

Reviewers: José Armando García Sancio <jsancio@apache.org>, Colin P. McCabe <cmccabe@apache.org>, Chia-Ping Tsai <chia7712@apache.org>

dajac committed:
KAFKA-17621; Reduce logging verbosity on ConsumerGroupHeartbeat path (#17288)

While running large scale performance tests, we noticed that the logging on the ConsumerGroupHeartbeat path took a significant amount of CPU. It is mainly due to the very large data structures that we print out. I made a pass on those logs and I switched some of them to debug.

Reviewers: Lianet Magrans <lianetmr@gmail.com>

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicIds.java:
```diff
@@ -21,25 +21,169 @@
+     *
+     * Provides an implementation of equals and hashCode based on the underlying TopicsImage.
+     */
+    public abstract static class BaseTopicResolver implements TopicResolver {
```
nit: Should we keep it private?

squah-confluent replied:
Yes, it doesn't need to be public.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicIds.java:
```diff
@@ -164,20 +308,20 @@ public boolean equals(Object o) {
     public String toString() {
         return "TopicIds(topicNames=" + topicNames +
-            ", image=" + image +
+            ", resolver=" + resolver +
```
nit: If we toString it here, should we add toString methods to the resolvers too?

squah-confluent replied:
I'll add a toString implementation that returns `<class name>(image=<image>)`.

dajac reviewed (commented):
@squah-confluent Thanks for the patch. I left a few nits.

xijiu committed:
KAFKA-17563 Move `RequestConvertToJson` to server module (#17223)

Reviewers: Chia-Ping Tsai <chia7712@apache.org>

FrankYang0529 committed:
KAFKA-17154: New consumer subscribe may join group without a call to consumer.poll (#17165)

Reviewers: Lianet Magrans <lmagrans@confluent.io>, TaiJuWu <tjwu1217@gmail.com>

FrankYang0529 committed:
KAFKA-16683 Extract security-related helpers from scala.TestUtils to java class (#16912)

Reviewers: Chia-Ping Tsai <chia7712@apache.org>

lianetm committed:
MINOR: Remove duplicate method (#17291)

Reviewers: Chia-Ping Tsai <chia7712@apache.org>

rreddy-22 committed:
KAFKA-17277: [2/2] Add feature dependency command to the storage and the feature command tool (#17093)

This patch belongs to the ongoing efforts of implementing KIP-1022.

Added feature-dependencies command to look up dependencies for a given feature version supplied by --feature flag. If the feature is not known or the version not yet defined, we throw an error.

Examples :
bin/kafka-feature feature-dependencies --feature transaction.version=2
transaction.version=2 requires:
metadata.version=4 (3.3-IV0) (listing any other version dependencies)

bin/kafka-feature feature-dependencies --feature metadata.version=17
metadata.version=17 (3.7-IV2) has no dependencies

Reviewers: Justine Olshan <jolshan@confluent.io>, Artem Livshits <alivshits@confluent.io>

fonsdant committed:
KAFKA-17488: Cleanup (test) code for Kafka Streams "metric version" (#17182)

This PR simply StreamsMetricsImpl to avoid passing in the unused "metric version" parameter.

Reviewers: Matthias J. Sax <matthias@confluent.io>

mjsax committed:
KAFKA-16331: remove KafkaClientSupplier from StreamsProducer (#17259)

With EOSv1 removed, we don't need to create a producer per task, and thus can simplify the code by removing KafkaClientSupplier from the deeply nested StreamsProducer, to simplify the code.

Reviewers: Bill Bejeck <bill@confluent.io>

mumrah committed:
KAFKA-17628 New workflows for automating run approvals (#17290)

This patch allows workflow runs to be controlled by the ci-approved label on Pull Requests. Rather than manually approving each workflow run explicitly, committers can now add the appropriate label and the new "CI Requested" and "PR Labeled" workflows will auto-approve the requested run.

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

apoorvmittal10 committed:
KAFKA-17626: Move common fetch related classes from storage to server-common (#17289)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>, David Arthur <mumrah@gmail.com>

dongnuo123 committed:
MINOR: Set default group.consumer.migration.policy to BIDIRECTIONAL (#17295)

This is a small patch to change the default value of group.consumer.migration.policy to BIDIRECTIONAL.

Reviewers: David Jacot <djacot@confluent.io>

dimitarndimitrov committed:
MINOR: Fix a race and add JMH bench for HdrHistogram (#17221)

FrankYang0529 committed:
KAFKA-17581: AsyncKafkaConsumer can't unsubscribe invalid topics (#17244)

Reviewers: Kirk True <ktrue@confluent.io>, Lianet Magrans <lmagrans@confluent.io>

FrankYang0529 committed:
KAFKA-17619: Remove zk type and instance from ClusterTest (#17284)

Reviewers: Colin P. McCabe <cmccabe@apache.org>, Chia-Ping Tsai <chia7712@gmail.com>, David Arthur <mumrah@gmail.com>

OmniaGM committed:
KAFKA-14572: Migrate EmbeddedKafkaCluster used by Streams integration tests from EmbeddedZookeeper to KRaft (#17016)

Migrate the EmbeddedKafkaCluster from the EmbeddedZookeeper to KRaft

Reviewers Bill Bejeck <bbejeck@apache.org>

mumrah committed:
KAFKA-17636 Fix missing SCRAM bootstrap records (#17305)

Fixes a regression introduced by #16669 which inadvertently stopped processing SCRAM arguments from kafka-storage.sh

Reviewers: Colin P. McCabe <cmccabe@apache.org>, Federico Valeri <fedevaleri@gmail.com>

mumrah committed:
MINOR Don't run tests for Draft PRs (#17299)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

devanshikhatsuriya committed:
KAFKA-6197: Update Streams API and Javadoc references in documentation (#17215)

Reviewers: Matthias J. Sax <matthias@confluent.io>

squah-confluent committed:
fixup: add toString implementation to TopicResolvers

squah-confluent committed:
fixup: make BaseTopicResolver private

squah-confluent committed:
Revert bad merge

This reverts commit 1f9c7ae9cbbe2b06b620ad7d4c5164bb8c84cb2d.
This reverts commit 0b97fa6bceb0d530f0c34adcda664f94affa46e3.
This reverts commit d9254d571e4c0f72fe14cf396fa912130af42f7c.
This reverts commit b905a9f9418f03f822594a1864845024d3c6963b.
This reverts commit f438bb441e625a98e80d9348dbdcad7b12c22677.
This reverts commit f0a2d4332972363b9765aab7da317fa7c413c435.
This reverts commit 945023da77dbccfd3cb89536d1911abac808f70d.
This reverts commit 2df3ff09f9ee3762e186d2e85ca0d17f04d2a32d.
This reverts commit 3b555b8f27155c5e851e45051695817de5fad0f6.
This reverts commit 2059889c13b7d5c8cb70b65a0f312fcaeeb37719.
This reverts commit fe5056eb502b70c5cca760549b4b69134079e575.
This reverts commit 0589e935f88fbcd9b3f65629676c8b89460185d0.
This reverts commit 99e9f81aef2c52d646967315ddc77ccea5063fca.
This reverts commit 813d79c4349b6c94dda65fe089dca842b17b79c9.
This reverts commit 5aaf623ac11438bcd74179f39890903b0d58530c.
This reverts commit 0d8485a135a09dd858158287f52c587a7030e47a.
This reverts commit 25b048e44cadb795369842584fd960462449a42e.
This reverts commit 536dfbbafd36e13b87457b22b664e1bc5b383e8e.
This reverts commit 655cddd4bf6be1419fa1cb6fd46490da351f155e.
This reverts commit 48c022cd05f87358c38fc91c3e9bacbbe34afaef.
This reverts commit 03b4cdc9c2cf7c49aa36afaec2873527b6baa81d.
This reverts commit a59e8c51ba44428b174e65f24f21c0f15a1fbd98.

squah-confluent committed:
Merge remote-tracking branch 'origin/trunk' into squah-cache-topicids-topic-resolution

squah-confluent committed:
Merge pull request #135 from confluentinc/squah-cache-topicids-topic-resolution

Updates for https://github.com/apache/kafka/pull/17285

mumrah commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/TopicIds.java:
```diff
@@ -21,25 +21,175 @@
+     * Provides an implementation of equals, hashCode and toString based on the underlying
+     * TopicsImage.
+     */
+    private abstract static class BaseTopicResolver implements TopicResolver {
```
Why do we need this abstract class? Can we just have the interface and concrete class?

squah-confluent replied:
I can remove it.

mumrah reviewed (commented):
@squah-confluent, Seeing that the cache is not actually used outside of tests and benchmarks, I'm guessing this is still WIP.

Does this come down to performance differences between HashMap and PCollectionsImmutableMap?

If we decide we really need faster topic ID to name lookups, I would consider adding it to TopicsImage. Managing a cache outside of the image will be a bit difficult. 

Another thing to consider is the lifetime of the cache. Do we really need the ID + name mappings kept in memory forever?

dajac commented:
@mumrah 

> Another thing to consider is the lifetime of the cache. Do we really need the ID + name mappings kept in memory forever?

The lifetime of the cache is bound to the call. It is not kept forever.

> Does this come down to performance differences between HashMap and PCollectionsImmutableMap?

Yes.

> If we decide we really need faster topic ID to name lookups, I would consider adding it to TopicsImage. Managing a cache outside of the image will be a bit difficult.

We could consider this separately. At the moment, we don't really have the time to do it. The current strategy seems to be a good tradeoff at the moment given that it is only bound to the call and not kept forever.

mumrah commented:
@dajac thanks for the explanation, makes sense. Can we include a javadoc on the class describing the expected lifetime of this class?

squah-confluent committed:
fixup: remove AbstractTopicResolver, inline equals and hashCode into TopicIds

squah-confluent committed:
fixup: add comment about lifetime of CachedTopicResolver

squah-confluent commented:
@mumrah 
> Seeing that the cache is not actually used outside of tests and benchmarks, I'm guessing this is still WIP.
It's used in TargetAssignmentBuilder.build(), which is used by the new group coordinator.

I've updated the javadoc to describe the lifetime of the cache.

dajac reviewed (approved):
LGTM, thanks.

dajac commented:
@mumrah I will merge it. If you have further comments, @squah-confluent can address separately.
